terraform/mysql/provision/main.tf

```diff
@@ -50,7 +50,7 @@ resource "random_password" "password" {
 resource "aws_db_instance" "db_instance" {
   allocated_storage                     = var.storage_gb
   storage_type                          = var.storage_type
-  iops                                  = var.storage_type == "io1" ? var.iops : null
```

Comment:
I don't like this behaviour. I know it was the existing functionality and not something you added in this commit.

Basically, we are **"removing"** the `iops` value a customer might be passing if the `storage_type` is not one of the `valid_storage_types_for_iops` (previously only `io1`).

From a UX experience this is disconcerting.
A customer may ask:
> Why is my iops sometimes not having an effect?

We shouldn't use the strategy of **"removing"** values lightly.

Instead, we should:
- Use a pre/postcondition and raise an error if an invalid combination is used
- Let terraform continue execution with an invalid combination, and ensure the customer get's a reasonable error.

We need to be very careful with default values in these cases.

1. Imagine a customer gets the message `invalid combination gp2 with iops 3000`.
2. The customer review the command they ran but don't see any reference to `iops`.
3. The customer don't realise that it is caused by a default value for `iops`.
4. The customer contacts us.

-------------
Another bug or misbehaviour caused by silently modifying or removing input variables is:
1. Imagine an operator defines a plan with default `iops: 40000`
2. Imagine the operator sets `storage_type: null` to allow consumers to choose the one they prefer.
3. Suppose the operator forgets to limit `storage_type` with an enum. Or simply thinks (you know what? it shouldn't matter, consumers can't override the `iops` property so it should fail with other `storage_types`).
4. A consumer comes and sets `storage_type: standard` and it works 😱!!!!

Reply:
In fact, it just crossed my mind that this PR was needed because of this **"ignore"** logic.
If this logic wasn't in place, customers could have harnessed `iops` with `gp3` without issues.

Reply:
The ignore logic was implemented at some point to support  setting a default for iops. We might want to review this at some point. I do agree ignoring a setting under certain circumstances is not the happiest solution and that also causes us to actively maintain a list of values. I would leave this as a future enhancement in the interest of time.